### PR TITLE
Leaf/a5.1 input change education

### PR DESCRIPTION
**Pull request by Seisigh**
#  Features
- Allows input from the user to insert education certification, education description, and school. 
- Allows multiple instances of education input and edit with an auto-suggestion.

# Implementation 

### Group Accordion
- Created reusable accordion group components which takes in arguments for its initial list of accordions.
- Within this group of accordions, each accordion has two combo boxes that allow for both custom user inputs and selection from suggested lists of options.
- There is also a description text field for user input under a drop-down.
- The group accordion has a state of all the accordions' data.
- The data is an array of objects each containing the following properties: 
        - comboOneVal : a string that represents the current value of the first combo box of the corresponding accordion.
        - comboTwoVal : a string that represents the current value of the second combo box of the corresponding accordion.
        - descText: a string that represents the current value of the description text field of the corresponding accordion.
- Group accordion passes down data along with a method to change the data.
- Accordions pass down both functions and data for each input. The functions handle the user input for that input component and edit the corresponding data within the group accordion state.  

### Error
- We are able to pass in custom error validation functions for each input (takes in a string value[user input] and returns true if it satisfies the condition or a string that represents the error message.)
- If there are errors, they will be displayed in the error tooltip within the corresponding input component. 

### Add Accordion
- Clicking the add button will display a modal that prompts the user to enter data for a new accordion.
- The data is not saved until all are validated AND the user clicks on 'save' which will then be added to the data and displayed on the screen.
- If it is clicked off, the data will not be added but will be saved in the modal for reuse, but if successfully added, the modal's data will be wiped and will start fresh up on the next click of the add button.

### Types
- Created Education type which is an array containing objects that each contain a degree, school, and description.

# What you should see

The group accordion for education.
image_2023-10-04_220444419

The modal for adding new accordions.
image

Error for adding a new accordion.
image
image

Custom Input.
image

Added two new accordions with the add button.
image

### CODE REVIEW

**Comment by JRSiwiecki** [comment]
I haven't started a review yet, but from a brief look at the description and screenshots, shouldn't there be a way to input and display start and end dates?

**Comment by Darren52BT** [comment]
> I haven't started a review yet, but from a brief look at the description and screenshots, shouldn't there be a way to input and display start and end dates?

It does make a lot of sense to include that but I don't see anything talking about dates on the issue or in the figma. Perhaps that could be done in another issue?

**Comment by JRSiwiecki** [comment]
> > I haven't started a review yet, but from a brief look at the description and screenshots, shouldn't there be a way to input and display start and end dates?
> 
> It does make a lot of sense to include that but I don't see anything talking about dates on the issue or in the figma. Perhaps that could be done in another issue?

Possibly. The start and end dates were added retroactively anyways, so another issue is probably fine. (Would either of you mind making the issue pllllllllleeeeeeeeeeeaseeeeeee 🥺)

**Comment by Darren52BT** [comment]
> > > I haven't started a review yet, but from a brief look at the description and screenshots, shouldn't there be a way to input and display start and end dates?
> > 
> > 
> > It does make a lot of sense to include that but I don't see anything talking about dates on the issue or in the figma. Perhaps that could be done in another issue?
> 
> Possibly. The start and end dates were added retroactively anyways, so another issue is probably fine. (Would either of you mind making the issue pllllllllleeeeeeeeeeeaseeeeeee 🥺)

done 👍 #315

File: meshapp/src/profile/profile-accordion-combobox.tsx [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      value: ele,
+      label: ele,
+    };
+  });
+
+  //onChange handler
+  const onChangeHandler = (
+    event: SyntheticEvent<Element, Event>,
+    val: string | option | null
+  ) => {
```
Comment (JRSiwiecki) [inline meshapp/src/profile/profile-accordion-combobox.tsx:73]: This comment is probably too redundant to exist.

File: meshapp/src/profile/profile-accordion.tsx [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+/**
+ * A React Component that returns an accordion with two comboboxes and a description textfield.
+ * Provides editing capabilities for these inputs.
+ *
+ * Used in Profile Group Accordion (src/profile/profile-group_accordion.tsx)
+ *
+ * @param props - Properties of the component
+ * @param {string} props.headerOne - value for combobox of this accordion's headerOne, passed down from group state array data
+ * @param {string} props.headerTwo - value for combobox of this accordion's headerTwo, passed down from group state array data
```
Comment (JRSiwiecki) [inline meshapp/src/profile/profile-accordion.tsx:24]: Can these names be more descriptive? Based on the comment I'm still not entirely sure what these headers actually represent.

File: meshapp/src/profile/profile-page.tsx [hunk 3]
```diff
@@ -290,7 +307,34 @@ const ProfileEducation = (props: any) => {
...
+          ]}
+          headerTwoOptions={["Cal Poly Pomona", "Mt. San Antonio College"]}
+          descPlaceholder="Enter a description here"
+          headerOneErrValidations={[
+            isEmpty("Level of Education"),
+            whiteSpace("Level of Education"),
+          ]}
+          headerTwoErrValidations={[isEmpty("School"), whiteSpace("School")]}
+          descErrValidations={[
+            isEmpty("Description"),
```
Comment (JRSiwiecki) [inline meshapp/src/profile/profile-page.tsx:333]: Do we really need a check for if the description is empty? I feel that this should be some optional thing to add the description, not necessarily be forced on the user.

**Review by JRSiwiecki — COMMENTED** [review]
Despite how large this issue is, I really don't have much to comment on because upon testing it seems like it was handled very well. The only things I mention are small things like optional descriptions and variable naming & comment conventions. 

This is all very good! Re-ping me on discord when either of you have a moment to review my bit of feedback. 🥇

P.S. Also remember to using closing keywords along with the task # in the description of your PRs.

File: meshapp/src/profile/profile-accordion-combobox.tsx (hunk 1 above)
Comment (chrispybokchoy) [reply meshapp/src/profile/profile-accordion-combobox.tsx]: my latest commit #16 should resolve it.  Just removed comment "//onChange handler."

File: meshapp/src/profile/profile-page.tsx (hunk 3 above)
Comment (chrispybokchoy) [reply meshapp/src/profile/profile-page.tsx]: removed "isEmpty("description")" so now it should not check for empty description;  description can now be empty (commit #16)

File: meshapp/src/profile/profile-accordion.tsx (hunk 2 above)
Comment (chrispybokchoy) [reply meshapp/src/profile/profile-accordion.tsx]: Darren changed the variable names (commit #15) headerOne -> ComboOneVal and headerTwo -> ComboTwoVal.

**Review by JRSiwiecki — APPROVED** [review]
Good work!

**Review by misslame — APPROVED** [review]
Beautiful code. I have literally no comments. Well done. Error handling, breaking up the code in sections, using docs, I am nothing but proud

**Pull request merged**